GH-73991: Add pathlib.Path.move that can handle rename across FS

oz123 opened the pull request:
With this change, ``Pathlib.move`` adds the ability
to handle renaming across file system and also preserve metadata
when renaming, since ``shutil.move`` using ``shutil.copy2`` is used
under the hood.

https://bugs.python.org/issue46317

<!-- issue-number: [bpo-46317](https://bugs.python.org/issue46317) -->
https://bugs.python.org/issue46317
<!-- /issue-number -->


<!-- gh-issue-number: gh-73991 -->
* Issue: gh-73991
<!-- /gh-issue-number -->

barneygale commented on Doc/library/pathlib.rst:
```diff
@@ -1034,6 +1034,14 @@ call fails (for example because the path doesn't exist).
+   .. note::
+      The order of arguments (link, target) is the reverse
+      of :func:`os.symlink`'s.
+
```
Accidental copy-paste?

oz123 replied:
Yep. Removed.

barneygale commented on Doc/library/pathlib.rst:
```diff
@@ -1034,6 +1034,14 @@ call fails (for example because the path doesn't exist).
+      of :func:`os.symlink`'s.
+
+   .. note::
+      This method can not move files from one file system to another.
```
"cannot" or "can't"

oz123 replied:
Change this to conform with the rest of the docs (using can't).

barneygale commented on Doc/library/pathlib.rst:
```diff
@@ -1034,6 +1034,14 @@ call fails (for example because the path doesn't exist).
+      of :func:`os.symlink`'s.
+
+   .. note::
+      This method can not move files from one file system to another.
```
"filesystem" (without a space) seems to be the usual spelling in the pathlib docs.

oz123 replied:
Fixed.

barneygale commented on Doc/library/pathlib.rst:
```diff
@@ -1048,9 +1056,25 @@ call fails (for example because the path doesn't exist).
+
+   Rename this file or directory to the given *target*, and return a new Path
+   instance pointing to *target*.  If *target* points to an existing file or
+   directory, it will be unconditionally replaced.
```
This doesn't agree with the `shutil.move()` docs:

> If the destination is an existing directory, then src is moved inside that directory

oz123 replied:
Fixed. Thanks for that!

barneygale commented on Lib/pathlib.py:
```diff
@@ -1245,6 +1247,21 @@ def replace(self, target):
+        If the destination is on the current filesystem, then os.rename() is used.
+        Otherwise, src is copied to dst using copy_function and then removed.
+        In case of symlinks, a new symlink pointing to the target of will be
+        created in or as target and the original path will be removed.
```
Can you re-word this sentence? I don't follow it I'm afraid.

oz123 replied:
See my rephrasing. Maybe it's better now.

barneygale commented on Lib/test/test_pathlib.py:
```diff
@@ -2026,6 +2026,23 @@ def test_replace(self):
+        self.assertEqual(replaced_q, self.cls(r))
+        self.assertEqual(os.stat(r).st_size, size)
+        self.assertFileNotFound(q.stat)
+
```
Suggestions for tests:

- Also test moving a file into an existing _directory_
- Also test symlinks
- Also mock out `os.rename()` so that you hit the copying code

oz123 replied:
@barneygale I added a test  for moving a file into an existing directory
The other tests will follow soon.

oz123 replied:
Mock `os.rename()` added.

oz123 replied:
Also added a test for symlink.

barneygale reviewed (commented):
I'd prefer `metadata=True` over `copy_function=shutil.copy2`, and would like to request a third opinion :-)

oz123 committed:
bpo-46317: Add pathlib.Path.move that can handle renaming across FS

With this change, ``pathlib.Path.move`` adds the ability
to handle renaming across file system and also preserve metadata
when renaming, since ``shutil.move`` using ``shutil.copy2`` is used
under the hood.

oz123 committed:
Fix grammar: replace wrong an with a

barneygale commented on Lib/pathlib.py:
```diff
@@ -1177,6 +1178,19 @@ def replace(self, target):
+        Returns the new Path instance pointing to the target path.
+        """
+        shutil.move(self, target, copy_function)
+        return self.__class__(target)
```
I think it would be better to return the actual destination of the file, i.e. the return value of `shutil.move()`,

oz123 replied:
Would it not violate the API of all the module? The methods `rename` and `replace`  wrap the target paths with `Path`.

barneygale replied:
Ah sorry, and wrap it in a Path:

```
return self.__class__(shutil.move(self, target, copy_function))
```

(this is _not_ the same as *target* - see the implementation of `move()`!)

oz123 replied:
Done! Thanks for that sharp eye!

MaxwellDupre reviewed (approved):
Ran 462 tests in 3.520s
OK (skipped=137)
Added test passed. Looks ok.
